src/jdk.jlink/share/classes/jdk/tools/jlink/internal/runtimelink/ResourceDiff.java

```diff
@@ -0,0 +1,290 @@
+            }
+            switch (kind) {
+            case ADDED:
+                {
+                    break; // null bytes for added is OK.
+                }
+            case MODIFIED: // fall-through
+            case REMOVED:
+                {
+                    if (resourceBytes == null) {
+                        throw new IllegalStateException("Original bytes needed for MODIFIED, REMOVED!");
+                    }
+                    break;
+                }
+            default:
+                break;
+            }
+            return new ResourceDiff(kind, name, resourceBytes);
+        }
+    }
+
+    /**
+     * Writes a list of resource diffs to an output stream
+     *
+     * @param diffs The list of resource diffs to write.
+     * @param out The stream to write the serialized bytes to.
+     */
+    public static void write(List<ResourceDiff> diffs, OutputStream out) throws IOException {
+        /*
+         * Simple binary format:
+         *
+         * <header>|<items>
+         *
+         * ****************************************
+         * HEADER info
+         * ****************************************
+         *
+         * where <header> is ('|' separation for clarity):
+         *
+         *  <int>|<int>
+         *
+         * The first integer is the MAGIC, 0xabba. The second integer is the
+         * total number of items.
+         *
+         * *****************************************
+         * ITEMS info
+         * *****************************************
+         *
+         * Each <item> consists of ('|' separation for clarity):
+         *
+         * <short>|<int>|<name-bytes-utf>|<int>|<resource-bytes>
+         *
+         * Where the individual items are:
+         *
+         * <short>:
+         *     The value of the respective ResourceDiff.Kind.
+         * <int>:
+         *     The length of the name bytes (in UTF-8).
+         * <name-bytes-utf>:
+         *     The resource name bytes in UTF-8.
+         * <int>:
+         *     The length of the resource bytes. 0 (zero) if no resource bytes.
+         *     A.k.a 'null'.
+         * <resource-bytes>:
+         *     The bytes of the resource as stored in the jmod files.
+         */
+        try (DataOutputStream dataOut = new DataOutputStream(out)) {
+            dataOut.writeInt(MAGIC);
+            dataOut.writeInt(diffs.size());
+            for (ResourceDiff d: diffs) {
+                dataOut.writeShort(d.kind.value());
+                byte[] buf = d.name.getBytes(StandardCharsets.UTF_8);
+                dataOut.writeInt(buf.length);
+                dataOut.write(buf);
+                buf = d.resourceBytes;
+                dataOut.writeInt(buf == null ? 0 : buf.length);
+                if (buf != null) {
+                    dataOut.write(buf);
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
Is this IOException swallowed by jlink when not running with the debugging option? I'm wondering about the printStackTrace here.

Reply:
I think this is a left-over of some devevelopment work. Removed now.